pkg/daemon/sync.go

```diff
@@ -41,67 +44,103 @@ type changeSet struct {
 }
 
 // Sync starts the synchronization of the cluster with git.
-func (d *Daemon) Sync(ctx context.Context, started time.Time, newRevision string, ratchet revisionRatchet) error {
-	// Make a read-only clone used for this sync
+func (d *Daemon) Sync(ctx context.Context, started time.Time, newRevision string, ratchet ratchet) error {
+	currentRevision, err := ratchet.CurrentRevision(ctx)
+	if err != nil {
+		return err
+	}
+
 	ctxt, cancel := context.WithTimeout(ctx, d.GitTimeout)
-	working, err := d.Repo.Export(ctxt, newRevision)
+	defer cancel()
+
+	cleanupClone := func(clone *git.Export) {
+		if err := clone.Clean(); err != nil {
+			d.Logger.Log("error", fmt.Sprintf("cannot clean clone: %s", err))
+		}
+	}
+
+	// Get last-synced resources for comparison
+	lastResources, err := ratchet.CurrentResources(ctx)
 	if err != nil {
 		return err
 	}
-	cancel()
-	defer func() {
-		if err := working.Clean(); err != nil {
-			d.Logger.Log("error", fmt.Sprintf("cannot clean sync clone: %s", err))
+	// Resource map is not initialized
+	if lastResources == nil {
+		if currentRevision == "" {
+			// Initial clone
+			lastResources = make(map[string]resource.Resource)
+		} else {
+			// Clone last-synced revision
```

Comment:
Are there circumstances in which a revision was set, but not the resources associated with that revision? If there are, I think it's worth enumerating them here to justify this special case. I would expect the revision and resources to be set atomically -- either both, or neither.

Reply:
Thank you. You are right.
The revision and resources should be set atomically.

But if I understand correctly, there is a special case -- in a test that uses `daemon.WithWorkingClone()`, only revision is set and resource map is left uninitialized.
I think a proper solution for addressing this special case is tweaking test code to make them not a special case, but I do not have enough experience with Flux code to do that.
Instead, I added a comment in `Daemon.Sync()` and made it emit a warning when entering the special case.

Reply:
Now I look into those tests, I realise that the revision comes from the sync state -- either a tag, or an annotation on the secret -- so most of the time when the loop _starts_, it will have a current revision without having a corresponding set of resources. Sorry, I got that wrong in my review.

So, a changed assessment: I think if you factor a method out for getting the resources for a revision, this could be a bit tidier. (Neither of the exported git repos are used after calculating the manifests, so the method could create and tidy those up, I think.)

Reply:
Ah, I re-read `daemon` package and I realized that, as you said, the sync loop starts with the "current revision" being set most of the time, not only in tests.

I have factored out some methods for loading `manifests.Store` and last-synced resources.